processing/src/main/java/io/druid/query/search/SearchQueryRunner.java

```diff
@@ -104,47 +108,63 @@ public SearchQueryRunner(Segment segment)
         dimsToSearch = dimensions;
       }
 
-      final BitmapFactory bitmapFactory = index.getBitmapFactoryForDimensions();
-
-      final ImmutableBitmap baseFilter =
-          filter == null ? null : filter.getBitmapIndex(new ColumnSelectorBitmapIndexSelector(bitmapFactory, index));
-
+      boolean useStorageAdapter = false;
       for (DimensionSpec dimension : dimsToSearch) {
         final Column column = index.getColumn(dimension.getDimension());
         if (column == null) {
           continue;
         }
-
-        final BitmapIndex bitmapIndex = column.getBitmapIndex();
-        ExtractionFn extractionFn = dimension.getExtractionFn();
-        if (extractionFn == null) {
-          extractionFn = IdentityExtractionFn.getInstance();
+        if (!column.getCapabilities().hasBitmapIndexes()) {
+          // If one of the columns lacks bitmap indexes, fallback to using StorageAdapter and do a per-row scan.
+          // Could optimize this by using bitmaps when available, full scan otherwise
+          useStorageAdapter = true;
+          break;
         }
-        if (bitmapIndex != null) {
-          for (int i = 0; i < bitmapIndex.getCardinality(); ++i) {
-            String dimVal = Strings.nullToEmpty(extractionFn.apply(bitmapIndex.getValue(i)));
-            if (!searchQuerySpec.accept(dimVal)) {
-              continue;
-            }
-            ImmutableBitmap bitmap = bitmapIndex.getBitmap(i);
-            if (baseFilter != null) {
-              bitmap = bitmapFactory.intersection(Arrays.asList(baseFilter, bitmap));
-            }
-            if (bitmap.size() > 0) {
-              MutableInt counter = new MutableInt(bitmap.size());
-              MutableInt prev = retVal.put(new SearchHit(dimension.getOutputName(), dimVal), counter);
-              if (prev != null) {
-                counter.add(prev.intValue());
+      }
+
+      if (!useStorageAdapter) {
+        final BitmapFactory bitmapFactory = index.getBitmapFactoryForDimensions();
+        final Map<String, DimensionColumnReader> readers = index.getDimensionReaders();
+        final ImmutableBitmap baseFilter =
```

Comment:
Looking at this code has got me wondering.  If `baseFilter` is `null`, why are we doing anything with the bitmaps in the first place?  We can just return all the values that match the predicate and be done with it.

Also, if the `baseFilter` is empty, we can short-circuit all the logic and return early.